test/plugin/test_in_tail.rb

```diff
@@ -2856,17 +2872,26 @@ def test_updateTW_after_refreshTW
         Fluent::FileWrapper.open("#{@tmp_dir}/tail.txt", "wb") {|f| f.puts "file2 log1"}
 
         # This reproduces the following situation:
-        #     Rotation => refresh_watchers => update_watcher
+        #   Rotation => refresh_watchers => update_watcher
         # This add a new TailWatcher: TailWatcher(path: "tail.txt", inode: inode_1)
-        #     This overwrites `@tails["tail.txt"]`
+        #   @tails => {
+        #     inode_0: TailWatcher(path: "tail.txt", inode: inode_0),
+        #     inode_1: TailWatcher(path: "tail.txt", inode: inode_1),
+        #   }
         d.instance.refresh_watchers
 
-        # `watch_timer` calls `TailWatcher::on_notify`, and then `update_watcher` updates the TailWatcher:
+        # `watch_timer` calls `TailWatcher::on_notify`, and then `update_watcher` trys to update the TailWatcher:
         #     TailWatcher(path: "tail.txt", inode: inode_0) => TailWatcher(path: "tail.txt", inode: inode_1)
-        # The old TailWathcer is detached here since `rotate_wait` is just `1s`.
+        # However, it is already added in `refresh_watcher`, so this only detach the old TailWatcher.
+        # It is detached here since `rotate_wait` is just `1s`.
+        #   @tails => {
+        #     inode_0: TailWatcher(path: "tail.txt", inode: inode_0) **DETACHED**,
+        #     inode_1: TailWatcher(path: "tail.txt", inode: inode_1),
+        #   }
         sleep 3
 
-        # This adds a new TailWatcher: TailWatcher(path: "tail.txt1", inode: inode_0)
+        # This does nothing because inode_0 still exists in @tails. (The TailWatcher is detached, but it is not removed.)
+        # TODO: Check if this is expected.
         d.instance.refresh_watchers
```

Comment:
I think we need to figure out if this is an expected behavior.

Reply:
# Master

* `update_watcher` overwrites the path of `@tails`, so the old inode disappears from `@tails`.
* So, `refresh_watcher` find the old inode as a new target.
  * It should not cause log duplication. PositionFile should remember the pos.
  * Related to #4221

```rb
@tails => {
  tail.txt: TailWatcher(path: "tail.txt", inode: inode_1),
  tail.txt1: TailWatcher(path: "tail.txt1", inode: inode_0) **Newly added**,
}
```

# This PR

* `update_watcher` doesn't overwrites `@tails`, so the old inode still exists in `@tails`.
* So, `refresh_watcher` can't find the old inode as a new target.

```rb
@tails => {
  inode_0: TailWatcher(path: "tail.txt", inode: inode_0) **DETACHED**,
  inode_1: TailWatcher(path: "tail.txt", inode: inode_1),
}
```

# Question

I'm confusing what the expected behavior is...

Reply:
I think the following must be ensured when an inode is rotated.
We need to add some fixes for this.

* Remove the TailWatcher tied to the old path
  * Remove it from `@tails` and detach it
  * Unwatching has nothing to do with the rotation. (When `refresh_watcher` can't find the inode, then it should be unwatched)
* Create a new TailWathcer tied to the new path.

Reply:
What makes this complicated is that we want to think about rotation based on one inode, but `update_watcher` is called based on one path...

Reply:
I think we need to do `@tails.delete(tail_watcher.ino)` even if `refresh_watcher` find the new file before.

https://github.com/fluent/fluentd/blob/155bdcabfc61180c0880a47725952e1db5fea49a/lib/fluent/plugin/in_tail.rb#L523-L529

Reply:
Done as https://github.com/fluent/fluentd/compare/155bdcabfc61180c0880a47725952e1db5fea49a..bc130634eebc03c45d8ab1170f8e5a162110d311